src/package.rs

```diff
@@ -429,73 +436,139 @@ impl FromStr for PackageType {
+    fn is_allowed(c: char) -> bool {
+        let is_ascii_lowercase_alphanumeric =
+            |c: char| c.is_ascii_alphanumeric() && !c.is_ascii_uppercase();
+        match c {
+            '-' => true,
+            c if is_ascii_lowercase_alphanumeric(c) => true,
+            _ => false,
+        }
+    }
 
-        ensure!(all_lower_alphanum, "invalid package name: {value} - only ASCII lowercase alphanumeric characters and dashes are accepted");
+    /// Validate a package name.
+    pub fn validate(name: &str) -> Result<(), PackageNameError> {
+        // validate length
+        if name.len() < 3 {
+            return Err(PackageNameError::Length(name.len()));
+        }
 
-        const UNEXPECTED_MSG: &str =
-            "unexpected error: package name length should be validated prior to first character check";
+        // validate first character
+        match name.chars().next() {
+            Some(c) if Self::is_allowed_start(c) => {}
+            Some(c) => return Err(PackageNameError::InvalidStart(c)),
+            None => unreachable!(),
+        }
 
-        ensure!(
-            value
-                .chars()
-                .next()
-                .ok_or(miette!(UNEXPECTED_MSG))?
-                .is_alphabetic(),
-            "package names must begin with an alphabetic letter"
-        );
+        // validate all characters
+        let illegal = name
+            .chars()
+            .enumerate()
+            .find(|(_, c)| !Self::is_allowed(*c));
+        if let Some((index, c)) = illegal {
+            return Err(PackageNameError::InvalidCharacter(c, index));
+        }
 
+        Ok(())
+    }
+
+    /// New package name from string.
+    pub fn new<S: Into<String>>(value: S) -> Result<Self, PackageNameError> {
+        let value = value.into();
+        Self::validate(&value)?;
         Ok(Self(value))
     }
 }
 
-impl TryFrom<&str> for PackageName {
-    type Error = miette::Report;
-
-    fn try_from(value: &str) -> Result<Self, Self::Error> {
-        Self::try_from(value.to_string())
-    }
+#[test]
+fn can_parse_package_name() {
+    assert_eq!(PackageName::new("abc"), Ok(PackageName("abc".into())));
+    assert_eq!(PackageName::new("a"), Err(PackageNameError::Length(1)));
+    assert_eq!(
+        PackageName::new("4abc"),
+        Err(PackageNameError::InvalidStart('4'))
+    );
+    assert_eq!(
+        PackageName::new("serde_typename"),
+        Err(PackageNameError::InvalidCharacter('_', 5))
+    );
 }
 
-impl TryFrom<&String> for PackageName {
-    type Error = miette::Report;
+impl TryFrom<String> for PackageName {
+    type Error = PackageNameError;
 
-    fn try_from(value: &String) -> Result<Self, Self::Error> {
-        Self::try_from(value.to_owned())
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::new(value)
     }
 }
 
 impl FromStr for PackageName {
-    type Err = miette::Report;
+    type Err = PackageNameError;
```

Comment:
```suggestion
    type Err = miette::Report;
```

Reply:
I *heavily* disagree with using `miette::Report` here. `miette` is an error *report* library. It helps you create reports for errors. Meaing: you can use it to convert `low-level` errors into something that is pretty-printable. This is not the place to use it, that is why this PR has removed it. 

The idea is: you use fine-grained errors in your application, and at some point before presenting it to the user, you can turn your error into a report for presentation.

At this point, all we care about is parsing `PackageName`s and returning a useful error type that tells you *exactly why* the parsing failed. Which means an `enum` with fields and an `std::error::Error` implementation.
Downstream, you are then free to turn this into an error *report* using `miette`. But if we have this useful error type, we should use it instead. A `FromStr` should never return a `miette::Report`, that is a big code smell (to me).

Reply:
Please see #95 and #103

Reply:
This implementation is public, meaning that you leak this to the outside world